src/js/libgeo/Prover.js

```diff
@@ -68,21 +74,108 @@ function incidentPC(p, c) {
             var erg = General.mult(c.matrix, p.homog);
             erg = General.mult(p.homog, erg);
             erg = CSNumber.abs(erg);
-            if (erg.value.real < 0.0000000000001) {
-                p.incidences.push(c.name);
-                c.incidences.push(p.name);
-            }
+            return (erg.value.real < 0.0000000000001);
         }
     };
 }
 
 function checkConjectures() {
+    var debug = false;
+    if (debug) console.log("conjectures", conjectures.length);
     if (conjectures.length === 0) return;
-    // TODO: we need some randomized proving here:
-    // move free elements and check conjectures a number of times.
-    // For now assume that all conjectures could be verified.
+
+    //backupGeo
+    stateArrays.prover.set(stateIn);
+    // proverTmp holds temporary good configurations
+    stateArrays.proverTmp.set(stateIn);
+
+    var nummoves = 3;
+
+    // filter free objects which are involved in conjectures
+    var involved;
+
+    var recalcInvolved = function() {
+        involved = [];
+        conjectures.forEach(function(con) {
+            var invs = con.getInvolved();
+            var incis;
+            invs = invs.forEach(function(el) {
+                if (involved.indexOf(el) < 0) {
+                    involved.push(el);
+                    // also add incidences of involved objects
+                    incis = findAllIncis(el, []);
+                    incis.forEach(function(i) {
+                        if (involved.indexOf(i) < 0) involved.push(i);
+                    });
+                }
+            });
+        });
+    };
+
+    // recursively find all incidences to an geo object
+    var findAllIncis = function(el, list) {
```

Comment:
I don't think *incidences* is the correct concept here. What we need is to know which elements we should move in order to check a given conjecture. So I'd rather traverse the `args` properties of our elements, i.e. the chain of *defining* elements. One possible counterexample is a `CircleMP`, were one would want to move the midpoint even though it is not incident with the circle. So I'd say the name and implementation of that function should change.

Since we identify all objects by their name, we might speed things up by using an obect as a hash map, instead of those `list.indexOf` calls all over the place.

Reply:
I agree, a `involved[el] = true` and `if(involved[el])` would be both more readable and faster.